hawkbit/software_updatable.go

```diff
@@ -132,6 +132,49 @@ func (su *SoftwareUpdatable) SetContextDependencies(deps ...*DependencyDescripti
 	return su.setProperty(suPropertyContextDependencies, su.status.ContextDependencies)
 }
 
+// MapLastOperation converts the OperationStatus into map interface
+// This enables the logging of last operation status with keys and values
+func MapLastOperation(LastOperationPtr *OperationStatus) map[string]interface{} {
+	lastOperationMap := make(map[string]interface{})
+	lastOperationMap[correlationIdParam] = LastOperationPtr.CorrelationID
+	lastOperationMap[statusParam] = LastOperationPtr.Status
+
+	if LastOperationPtr.SoftwareModule != nil {
+		lastOperationMap[softwareModuleParam] = map[string]interface{}{
+			softwareModuleNameParam:    LastOperationPtr.SoftwareModule.Name,
+			softwareModuleVersionParam: LastOperationPtr.SoftwareModule.Version,
+		}
+	}
+
+	if LastOperationPtr.Software != nil {
+		var softwareDependencies []map[string]interface{}
+		for _, values := range LastOperationPtr.Software {
+			dependency := map[string]interface{}{
+				softwareGroupParam:   values.Group,
+				softwareNameParam:    values.Name,
+				softwareVersionParam: values.Version,
+				softwareTypeParam:    values.Type,
+			}
+			softwareDependencies = append(softwareDependencies, dependency)
+		}
+		lastOperationMap[SoftwareParam] = softwareDependencies
+	}
+
+	if LastOperationPtr.Progress != nil {
+		lastOperationMap[progressParam] = *LastOperationPtr.Progress
+	}
+
+	if LastOperationPtr.Message != "" {
+		lastOperationMap[messageParam] = LastOperationPtr.Message
+	}
+
+	if LastOperationPtr.StatusCode != "" {
+		lastOperationMap[progressParam] = LastOperationPtr.StatusCode
```

Comment:
`progressParam` must be `statusCodeParam`.

This conversion to map seems like an unusual approach to me. I expected that the the Stringer interface will be implemented for the structures - https://go.dev/tour/methods/17